python/lsst/consdb/pqserver.py

```diff
@@ -19,29 +19,109 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-from typing import Any, Iterable
+from enum import Enum
+from typing import Annotated, Any, Optional
 
+import astropy
 import sqlalchemy
 import sqlalchemy.dialects.postgresql
-from flask import Flask, request
+from fastapi import Body, Depends, FastAPI, HTTPException, Path, Query, Request, status
+from fastapi.exceptions import RequestValidationError
+from fastapi.responses import JSONResponse
+from pydantic import BaseModel, Field, field_validator
 
 from .utils import setup_logging, setup_postgres
 
-OBS_TYPE_LIST = ["exposure", "visit1", "ccdexposure", "ccdvisit1"]
-DTYPE_LIST = ["bool", "int", "float", "str"]
 
-OBS_ID_COLNAME_LIST = ["ccdvisit_id", "visit_id", "ccdexposure_id", "exposure_id", "obs_id"]
+class ObsTypeEnum(str, Enum):
+    EXPOSURE = "exposure"
+    VISIT1 = "visit1"
+    CCD_EXPOSURE = "ccdexposure"
+    CCD_VISIT1 = "ccdvisit1"
+
+    @classmethod
+    def _missing_(cls, value):
+        """Makes the enum case-insensitive, see
+        https://docs.python.org/3/library/enum.html
+        """
+        value = value.lower()
+        for member in cls:
+            if member.value == value:
+                return member
+        return None
+
+
+ObservationIdType = int
+
+
+# This shenanigan makes flake8 recognize AllowedFlexTypeEnum as a type.
+AllowedFlexType = bool | int | float | str
+AllowedFlexTypeEnumBase = Enum(
+    "AllowedFlexTypeEnumBase", {t.__name__.upper(): t.__name__ for t in AllowedFlexType.__args__}
+)
+AllowedFlexTypeEnum = AllowedFlexTypeEnumBase
+
+
+def convert_to_flex_type(ty: AllowedFlexTypeEnum, v: str) -> AllowedFlexType:
+    """Converts a string containing a flex database value into the
+    appropriate type.
+
+    Raises
+    ======
+    ValueError if the conversion is invalid
+    """
+    if ty.value == "bool":  # Special case
+        return v.lower() in ("true", "t", "1")
+    m = [t for t in AllowedFlexType.__args__ if t.__name__ == ty.value]
+    assert len(m) == 1
+    return m[0](v)
+
+
+class ObsIdColname(str, Enum):
```

Comment:
What does it do when you subclass from both `str` and `Enum`?  I've never seen that before and I don't see anything about it in the docs.

Reply:
Don't know what I was thinking there, but I was going for StrEnum.